parlai/utils/conversations.py

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Utility methods for conversations format.
+"""
+import datetime
+import json
+import os
+import random
+
+
+class Metadata:
+    """
+    Utility class for conversation metadata.
+
+    Metadata should be saved at <datapath>.metadata.
+    """
+
+    def __init__(self, datapath):
+        self._load(datapath)
+
+    def _load(self, datapath):
+        metadata_path = self._get_path(datapath)
+        if not os.path.isfile(metadata_path):
+            raise RuntimeError(
+                f'Metadata at path {metadata_path} not found. '
+                'Double check your path.'
+            )
+
+        with open(metadata_path, 'rb') as f:
+            metadata = json.load(f)
+
+        self.datetime = metadata['date']
+        self.opt = metadata['opt']
+        self.self_chat = metadata['self_chat']
+        self.speaker_1 = metadata['speaker_1']
+        self.speaker_2 = metadata['speaker_2']
+        self.extra_data = {}
+        for k, v in metadata.items():
+            if k not in ['date', 'opt', 'speaker_1', 'speaker_2']:
+                self.extra_data[k] = v
+
+    def read(self):
+        """
+        Read the relevant metadata.
+        """
+        print(f'Saved at: {self.datetime}')
+        print(f'Self chat: {self.self_chat}')
+        print(f'Speaker 1: {self.speaker_1}')
+        print(f'Speaker 2: {self.speaker_2}')
+        print('Opt:')
+        for k, v in self.opt.items():
+            print(f'\t{k}: {v}')
+        for k, v in self.extra_data.items():
+            print(f'{k}: {v}')
+
+    @staticmethod
+    def _get_path(datapath):
+        fle, _ = os.path.splitext(datapath)
+        return fle + '.metadata'
+
+    @classmethod
+    def save_metadata(
+        cls,
+        datapath,
+        opt,
+        self_chat=False,
+        speaker_1='human',
+        speaker_2=None,
+        **kwargs,
+    ):
+        """
+        Dump conversation metadata to file.
+        """
+        metadata = {}
+        metadata['date'] = str(datetime.datetime.now())
```

Comment:
Other than opt, should we consider moving this meta data into the conversation itself? e.g. datetime might legitimately want to be controlled independently (say, as we output from a log).

Reply:
Metadata is supposed to be conversation**S** level and not conversation level. Datetime is written when the conversations logs are written. 

There can be other information included in each individual conversation, as keys in the dict other than 'dialog' (like 'persona' for instance).